### PR TITLE
fix execution on Windows

### LINKED ISSUE #125: Support for execution on Windows
Since most of the implementation of this is platform independent (pulling images from repositories and extracting them), it should be possible to execute this on Windows.  Currently the main problem is that paths being fed into the analyzers are in Windows form (i.e. backslash), while the analyzers are looking for key files using UNIX path separators.

Need to either have the analyzers look for paths using the native separator, or have the paths fed to the analyzers consistently use UNIX path separators.

### PR DESCRIPTION
**Pull request by ronaudinho**
resolves #125

fix `filepath.Clean` and `filepath.Join` on Windows

manually tested for:
- [x] `fanal img`
- [x] `fanal ar`
- [x] `fanal fs`
- [x] `fanal repo`

### TODO
- [ ] fix disabled unit tests
  - [ ] [artifact/image/image_windows_test.go](https://github.com/ronaudinho/fanal/blob/b1d2216c212f20c675a3eaf7003169b648403b67/artifact/image/image_windows_test.go)
  - [ ] [artifact/local/fs_windows_test.go](https://github.com/ronaudinho/fanal/blob/b1d2216c212f20c675a3eaf7003169b648403b67/artifact/local/fs_windows_test.go)
  - [ ] [artifact/remote/git_test.go](https://github.com/ronaudinho/fanal/blob/b1d2216c212f20c675a3eaf7003169b648403b67/artifact/remote/git_windows_test.go)
  - [ ] [image/daemon/podman_test.go](https://github.com/ronaudinho/fanal/blob/b1d2216c212f20c675a3eaf7003169b648403b67/image/daemon/podman_test.go)
- [ ] fix integration tests for CI

#### Miscellaneous
- `fanal fs` and `fanal repo` yields different `ArtifactDetail.Applications` even though they share the same `Inspect` function

### CODE REVIEW

File: image/daemon/podman_test.go [hunk 1]
```diff
@@ -1,3 +1,5 @@
+// +build aix darwin dragonfly freebsd js,wasm linux netbsd openbsd solaris
```
Comment (ronaudinho) [inline image/daemon/podman_test.go:1]: disabled for CI purpose

**Comment by ronaudinho** [comment]
should we move **TODO** to its own issues? @knqyf263

File: walker/walk.go [hunk 2]
```diff
@@ -19,6 +20,7 @@ type WalkFunc func(filePath string, info os.FileInfo, opener analyzer.Opener) er
 
 func isIgnored(filePath string) bool {
 	filePath = strings.TrimLeft(filePath, "/")
+	filePath = filepath.Clean(filePath)
```
Comment (knqyf263) [inline walker/walk.go:23]: What about inserting `filepath.ToSlash` before L22? In that case, we probably need to replace `utils.PathSeparator` with `/`.

```
	filePath = filepath.ToSlash(filePath)
        filePath = strings.TrimLeft(filePath, "/")
	filePath = filepath.Clean(filePath)
```

File: walker/walk_test.go [hunk 3]
```diff
@@ -9,19 +11,49 @@ import (
...
+		t.Run(tt.name, func(t *testing.T) {
+			if tt.want {
+				for _, fp := range tt.dirs {
+					assert.True(t, isIgnored(fp))
+				}
+			} else {
+				for _, fp := range tt.dirs {
+					assert.False(t, isIgnored(fp))
+				}
+			}
```
Comment (knqyf263) [inline walker/walk_test.go:56]:
```suggestion
			for _, fp := range tt.dirs {
				assert.Equal(t, tt.want, isIgnored(fp))
			}
```

File: walker/walk_test.go [hunk 4]
```diff
@@ -1,3 +1,5 @@
+// +build aix darwin dragonfly freebsd js,wasm linux netbsd openbsd solaris
```
Comment (knqyf263) [inline walker/walk_test.go:1]: The following tests look the same as Windows. So, we don't need this build tag, right? In other words, do these tests pass even on Windows?

File: walker/walk_windows_test.go [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+		{
+			name: "ignore dir in backslash separator",
+			dirs: []string{`foo\.git`, `foo\node_modules`},
+			want: true,
+		},
+		{
+			name: "not ignore dirs",
+			dirs: []string{"foo", "foo/bar"},
+			want: false,
+		},
```
Comment (knqyf263) [inline walker/walk_windows_test.go:46]: We should keep only the difference.

```suggestion
		{
			name: "ignore dir in backslash separator",
			dirs: []string{`foo\.git`, `foo\node_modules`},
			want: true,
		},

```

File: image/daemon/podman_test.go (hunk 1 above)
Comment (knqyf263) [inline image/daemon/podman_test.go:1]:
```suggestion
// +build !windows
```

**Comment by knqyf263** [comment]
> testing.go:915: TempDir RemoveAll cleanup: remove C:\Users\RUNNER~1\AppData\Local\Temp\TestFSCache_PutArtifact_happy_path562076382\001\fanal\fanal.db: The process cannot access the file because it is being used by another process.

The test is failing currently due to the above error. We may have forgotten to close the DB.

**Comment by ronaudinho** [comment]
> > testing.go:915: TempDir RemoveAll cleanup: remove C:\Users\RUNNER~1\AppData\Local\Temp\TestFSCache_PutArtifact_happy_path562076382\001\fanal\fanal.db: The process cannot access the file because it is being used by another process.
> 
> The test is failing currently due to the above error. We may have forgotten to close the DB.

This did not happen when I ran the test on my local. I'll try to repro and find out the cause.

**Pull request closed without merging**